documentation/nginx_config.md
# config

[Back to resource list](../README.md#resources)

## Actions

- `:create`
- `:delete`

## Properties

| Name                    | Type            | Default                        | Description                                                                            |
| ----------------------- | --------------- | ------------------------------ | -------------------------------------------------------------------------------------- |
| `config_file`           | String          | `/etc/nginx/nginx.conf`        | The full path to the Nginx server configuration on disk.                               |
| `conf_cookbook`         | String          | `nginx`                        | Which cookbook to use for the configuration template.                                  |
| `conf_template`         | String          | `nginx.conf.erb`               | Which template to use for the configuration file.                                      |
| `conf_variables`        | Hash            | `{}`                           | Additional variables to include in conf template.                                      |
| `port`                  | Integer, String | `80`                           | Port to listen on.                                                                     |
| `server_name`           | String          | `node['hostname']`             | Sets the server_name directive.                                                        |
| `owner`                 | String          | `root`                         | Nginx file/folder owner.                                                               |
| `group`                 | String          | `root`                         | Nginx run-as/file/folder group.                                                        |
| `mode`                  | String          | `0640`                         | Nginx configuration file mode.                                                         |
| `folder_mode`           | String          | `0750`                         | Nginx configuration folder mode.                                                       |
| `log_dir_mode`          | String          | `0755`                         | Nginx log directory mode.                                                              |
| `log_dir_owner`         | String          | `root`                         | Nginx log directory owner.                                                             |
| `log_dir_group`         | String          | `root`                         | Nginx log directory group.                                                             |
| `process_user`          | String          | `www-data` (Debian) or `nginx` | Nginx run-as user.                                                                     |
| `process_group`         | String          | `www-data` (Debian) or `nginx` | Nginx run-as group.                                                                    |
| `worker_processes`      | Integer, String | `auto`                         | The number of worker processes.                                                        |
| `worker_connections`    | Integer, String | `1_024`                        | The maximum number of simultaneous connections that can be opened by a worker process. |
| `sendfile`              | String          | `on`                           | Enables or disables the use of sendfile().                                             |
| `tcp_nopush`            | String          | `on`                           | Enables or disables the use of the TCP_CORK socket option on Linux.                    |
| `tcp_nodelay`           | String          | `on`                           | Enables or disables the use of the TCP_NODELAY option.                                 |
| `keepalive_timeout`     | Integer, String | `65`                           | Timeout during which a keep-alive client connection will stay open on the server side. |
| `types_hash_max_size`   | Integer, String | `2_048`                        | Sets the maximum size of the types hash tables.on Linux.                               |
| `default_site_enabled`  | True, False     | `true`                         | Whether or not the default site is enabled.                                            |
| `default_site_cookbook` | String          | `nginx`                        | Which cookbook to use for the default site template.                                   |
| `default_site_template` | String          | `default-site.erb`             | Which template to use for the default site.                                            |
| `template_helpers`      | String, Array   | `nil`                          | Additional helper modules to include in the default site and config template.          |

## Examples

```ruby
nginx_config 'nginx' do
  action :create
  notifies :reload, 'nginx_service[nginx]', :delayed
end
```

### Using custom log directory permissions

```ruby
nginx_config 'nginx' do
  log_dir_mode '0750'
  log_dir_owner 'nginx'
  log_dir_group 'nginx'
  action :create
  notifies :reload, 'nginx_service[nginx]', :delayed
end
```


documentation/resources/config.md
# config

Creates nginx config.

```ruby
nginx_config 'default' do
  conf_cookbook       String
  conf_template       String
  conf_variables      Hash
  group               String
  worker_processes    [Integer, String]
  worker_connections  [Integer, String]
  sendfile            String
  tcp_nopush          String
  tcp_nodelay         String
  keepalive_timeout   [Integer, String]
  types_hash_max_size [Integer, String]
end
```

## Actions

| Action  | Default  |
| ------- | -------- |
| :create | &#x2713; |

## Properties

### `conf_cookbook`

Which cookbook to use for the conf template.

| property       | value         |
| -------------- | ------------- |
| Type           | String        |
| Default        | `nginx`       |

### `conf_template`

Which template to use for the conf.

| property       | value            |
| -------------- | ---------------- |
| Type           | String, Array    |
| Default        | `nginx.conf.erb` |

### `conf_variables`

Additional variables to include in conf template.

This property is intended to allow specifying variables in
a template specified by the conf_template or conf_cookbook
properties. The variables that can be set in the default
template may all be set using the resource properties.

| property       | value |
| -------------- | ----- |
| Type           | Hash  |
| Default        | `{}`  |

### `group`

Nginx group, if different than user.

| property       | value                 |
| -------------- | --------------------- |
| Type           | String                |
| Default        | `lazy { nginx_user }` |

### `worker_processes`

The number of worker processes.

| property       | value             |
| -------------- | ----------------- |
| Type           | [Integer, String] |
| Default        | `auto`            |

### `worker_connections`

The maximum number of simultaneous connections that can be opened by a worker process.

| property       | value             |
| -------------- | ----------------- |
| Type           | [Integer, String] |
| Default        | `1_024`           |

### `sendfile`

Enables or disables the use of sendfile().

| property       | value   |
| -------------- | ------- |
| Type           | String  |
| Default        | `on`    |
| Allowed Values | on, off |

### `tcp_nopush`

Enables or disables the use of the TCP_CORK socket option on Linux.

| property       | value   |
| -------------- | ------- |
| Type           | String  |
| Default        | `on`    |
| Allowed Values | on, off |

### `tcp_nodelay`

Enables or disables the use of the TCP_NODELAY option.

| property       | value   |
| -------------- | ------- |
| Type           | String  |
| Default        | `on`    |
| Allowed Values | on, off |

### `keepalive_timeout`

Timeout during which a keep-alive client connection will stay open on the server side.

| property       | value             |
| -------------- | ----------------- |
| Type           | [Integer, String] |
| Default        | `65`              |

### `types_hash_max_size`

Sets the maximum size of the types hash tables.

| property       | value             |
| -------------- | ----------------- |
| Type           | [Integer, String] |
| Default        | `2_048`           |


CHANGELOG.md
# nginx Cookbook CHANGELOG

This file is used to list changes made in each version of the nginx cookbook.

Standardise files with files in sous-chefs/repo-management
Standardise files with files in sous-chefs/repo-management

## [12.3.3](https://github.com/sous-chefs/nginx/compare/v12.3.2...v12.3.3) (2026-08-16)


### Bug Fixes

* **ci:** Update workflows to use release pipeline ([#656](https://github.com/sous-chefs/nginx/issues/656)) ([436cf3e](https://github.com/sous-chefs/nginx/commit/436cf3eae721839ccb69b2a05682e6a197ff9e89))
* repair workflow actionlint failures ([#666](https://github.com/sous-chefs/nginx/issues/666)) ([883e199](https://github.com/sous-chefs/nginx/commit/883e19901699c8938cf110c0bf944848e1454aa7))
* RSpec issues ([#638](https://github.com/sous-chefs/nginx/issues/638)) ([1928862](https://github.com/sous-chefs/nginx/commit/1928862d0eeca9d732075fa8271206e354004ff4))

## [12.3.2](https://github.com/sous-chefs/nginx/compare/12.3.1...v12.3.2) (2025-10-15)


### Bug Fixes

* **ci:** Update workflows to use release pipeline ([#656](https://github.com/sous-chefs/nginx/issues/656)) ([436cf3e](https://github.com/sous-chefs/nginx/commit/436cf3eae721839ccb69b2a05682e6a197ff9e89))

## 12.3.0 - *2025-04-01*

* Correct trailing semicolon for blocks - [#643](https://github.com/sous-chefs/nginx/issues/643)
* Support multiple options - [#644](https://github.com/sous-chefs/nginx/issues/644)

## 12.2.13 - *2025-01-07*

* Convert all test files to proper InSpec format with controls and metadata
* Rename test files to use _spec.rb suffix
* Add inspec.yml files for all test suites
* Add dependencies between test suites where appropriate
* Update test structure for better organization and reporting

## 12.2.12 - *2025-01-07*

Fix Rspec issues
Update workflows

## 12.2.10 - *2024-07-10*

Standardise files with files in sous-chefs/repo-management

## 12.2.6 - *2023-12-27*

Remove Fedora Linux from test matrix

## 12.2.3 - *2023-08-20*

* Skip creation of the list directory resource if the directory already exists as it is managed elsewhere

## 12.2.0 - *2023-04-13*

* Fix installation on EL9 platforms
* Fix installation on newer Amazon Linux platforms

## 12.1.7 - *2023-04-04*

Standardise files with files in sous-chefs/repo-management

## 12.1.5 - *2023-04-01*

Standardise files with files in sous-chefs/repo-management

## 12.1.4 - *2023-03-15*

Standardise files with files in sous-chefs/repo-management

## 12.1.3 - *2023-02-18*

Standardise files with files in sous-chefs/repo-management

## 12.1.2 - *2023-02-15*

Standardise files with files in sous-chefs/repo-management

## 12.1.1 - *2022-12-08*

Standardise files with files in sous-chefs/repo-management

## 12.1.0 - *2022-08-30*

* Security fixes
  * Fix nginx running as root on Debian family
  * Fix webserver able to overwrite/delete config files
  * Add Inspec tests

## 12.0.12 - *2022-04-20*

Standardise files with files in sous-chefs/repo-management

## 12.0.11 - *2022-02-04*

* Remove delivery and move to calling RSpec directly via a reusable workflow
* Update tested platforms

## 12.0.10 - *2021-11-26*

* Fix CI Badge Link in README

## 12.0.9 - *2021-11-24*

* Change the unit tested Debian platform to bullseye

## 12.0.8 - *2021-11-23*

* Rename `config_dir` to `conf_dir` in nginx_site.md

## 12.0.7 - *2021-10-08*

* Correct default value for `config_dir` in nginx_site.md

## 12.0.6 - *2021-09-04*

* Fix repo helper incorrect version for SLES
* Set default user/group to `root` for Debian/Ubuntu platforms

## 12.0.5 - *2021-08-30*

* Standardise files with files in sous-chefs/repo-management
* Fix ChefSpec test

## 12.0.4 - *2021-06-01*

* Standardise files with files in sous-chefs/repo-management

## 12.0.3 - *2021-05-12*

* Only trigger ohai refresh on plugin installation

## 12.0.2 - *2021-05-11*

* Standardise files with files in sous-chefs/repo-management

## 12.0.1 - *2021-05-10*

* Refactor the default site template
  * Cast to Array where possible to remove type check logic.
  * Make upstream options key optional.
  * Add test case for overriding template.

## 12.0.0 - *2021-05-07*

* Enable unified_mode resource for Chef 17
* Remove dependency on the deprecated ohai cookbook

## 11.5.3 - *2021-04-20*

* Fix site template: remove unmatching rbrace in options

## 11.5.1 - *2021-03-29*

* Fix start/restart/reload delayed actions - [@bmhughes](https://github.com/bmhughes)

## 11.5.0 - *2021-03-26*

* Fix generating multiple actions from the service resource - [@bmhughes](https://github.com/bmhughes)
* Kitchen test with CentOS 8 stream - [@bmhughes](https://github.com/bmhughes)
* Refactor helpers - [@bmhughes](https://github.com/bmhughes)

## 11.4.0 - *2021-02-27*

* Refactor `ResourceHelpers` library to inherit properties from the calling resource
* Allow for inclusion of arbitrary template helper modules

## 11.3.1 - *2021-02-25*

* Fixup the default site template and update nginx_site doc to match

## 11.3.0 - *2021-02-11*

* Add seperate `process_*` properties for nginx process use and group to `nginx_config` resource (Issue #572)
  * Deprecate and alias the `user` property as the `process_*` properties supercede it

## 11.2.0 - *2021-01-20*

* Add file and folder mode overrides to `nginx_config` and `nginx_site`

## 11.1.1 - *2020-12-08*

* Fix site containing directory creation when nginx_config is not used

## 11.1.0 - *2020-12-07*

* Add `repo_train` property to nginx_install to select stable/mainline when installing from the nginx repo.
* Add `packages_versions` property to nginx_install to specify specific package versions.

## 11.0.0 - *2020-12-02*

* Refactor cookbook to remove nxensite/nxdissite
* Remove all configuration resources from nginx_install.
* Move all general configuration to nginx_config.
* Refactor nginx_site to remove use of scripts.
* Create nginx_service resource to manage nginx service.
* Remove passenger install option.

## 10.6.1 (2020-11-16)


resources/config.rb
#
# Cookbook:: nginx
# Resource:: config
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

unified_mode true

include Nginx::Cookbook::Helpers

deprecated_property_alias 'user',
                          'process_user',
                          'The user property was renamed process_user in the 11.3 release of this cookbook. Please update your cookbooks to use the new property name. This alias will be removed in the next major version.'

property :config_file, String,
          description: 'The full path to the Nginx server configuration on disk.',
          default: lazy { nginx_config_file }

property :conf_cookbook, String,
          description: 'Which cookbook to use for the conf template.',
          default: 'nginx'

property :conf_template, String,
          description: 'Which template to use for the conf.',
          default: 'nginx.conf.erb'

property :conf_variables, Hash,
          description: 'Additional variables to include in conf template.',
          default: {}

property :port, [Integer, String],
          description: 'Port to listen on.',
          coerce: proc { |v| v.is_a?(Integer) ? v.to_s : v },
          default: 80

property :server_name, String,
          description: 'Sets the server_name directive.',
          default: lazy { node['hostname'] }

property :owner, String,
          description: 'File/folder user',
          default: 'root'

property :group, String,
          description: 'File/folder group',
          default: 'root'

property :mode, String,
          description: 'File mode',
          default: '0640'

property :folder_mode, String,
          description: 'Folder mode',
          default: '0750'

property :log_dir_mode, String,
          description: 'Log directory mode',
          default: lazy { platform_family?('debian') ? '0755' : '0750' }

property :log_dir_owner, String,
          description: 'Log directory owner',
          default: lazy { owner }

property :log_dir_group, String,
          description: 'Log directory group',
          default: lazy { group }

property :process_user, String,
          description: 'Nginx process user',
          default: lazy { nginx_user }

property :process_group, String,
          description: 'Nginx process group',
          default: lazy { nginx_group }

property :worker_processes, [Integer, String],
          description: 'The number of worker processes.',
          coerce: proc { |v| v.is_a?(Integer) ? v.to_s : v },
          default: 'auto'

property :worker_connections, [Integer, String],
          description: 'The maximum number of simultaneous connections that can be opened by a worker process.',
          coerce: proc { |v| v.is_a?(Integer) ? v.to_s : v },
          default: 1_024

property :sendfile, String,
          description: 'Enables or disables the use of sendfile().',
          equal_to: %w(on off),
          default: 'on'

property :tcp_nopush, String,
          description: 'Enables or disables the use of the TCP_CORK socket option on Linux.',
          equal_to: %w(on off),
          default: 'on'

property :tcp_nodelay, String,
          description: 'Enables or disables the use of the TCP_NODELAY option.',
          equal_to: %w(on off),
          default: 'on'

property :keepalive_timeout, [Integer, String],
          description: 'Timeout during which a keep-alive client connection will stay open on the server side.',
          coerce: proc { |v| v.is_a?(Integer) ? v.to_s : v },
          default: 65

property :types_hash_max_size, [Integer, String],
          description: 'Sets the maximum size of the types hash tables.',
          coerce: proc { |v| v.is_a?(Integer) ? v.to_s : v },
          default: 2_048

property :default_site_enabled, [true, false],
          description: 'Whether or not the default site is enabled.',
          equal_to: [true, false],
          default: true

property :default_site_cookbook, String,
          description: 'Which cookbook to use for the default site template.',
          default: 'nginx'

property :default_site_template, String,
          description: 'Which template to use for the default site.',
          default: 'default-site.erb'

property :default_site_variables, Hash,
          description: 'Additional variables to include in default site template.',
          default: {}

property :template_helpers, [String, Array],
          description: 'Additional helper modules to include in the default site and config template',
          coerce: proc { |p| p.is_a?(Array) ? p : [p] }

action_class do
  include Nginx::Cookbook::ResourceHelpers

  def default_site_enabled?
    new_resource.default_site_enabled
  end
end

action :create do
  %w(conf.d conf.http.d).each do |leaf|
    directory ::File.join(nginx_dir, leaf) do
      owner new_resource.owner
      group new_resource.group
      mode new_resource.folder_mode
    end
  end

  directory nginx_log_dir do
    owner new_resource.log_dir_owner
    group new_resource.log_dir_group
    mode new_resource.log_dir_mode
  end

  %w(default.conf example_ssl.conf).each do |config|
    file ::File.join(nginx_dir, "conf.d/#{config}") do
      action :delete
    end
  end

  nginx_site 'default-site' do
    cookbook new_resource.default_site_cookbook
    template new_resource.default_site_template
    conf_dir nginx_config_site_dir

    owner new_resource.owner
    group new_resource.group
    mode new_resource.mode
    folder_mode new_resource.folder_mode

    template_helpers new_resource.template_helpers

    variables(
      nginx_log_dir: nginx_log_dir,
      port: new_resource.port,
      server_name: new_resource.server_name,
      default_root: default_root
    ).merge!(new_resource.default_site_variables)
  end if default_site_enabled?

  template new_resource.config_file do
    cookbook new_resource.conf_cookbook
    source   new_resource.conf_template

    owner new_resource.owner
    group new_resource.group
    mode new_resource.mode

    helpers(Nginx::Cookbook::TemplateHelpers)
    new_resource.template_helpers.each { |th| helpers(::Object.const_get(th)) } unless new_resource.template_helpers.nil?

    variables(
      nginx_dir: nginx_dir,
      nginx_log_dir: nginx_log_dir,
      process_user: new_resource.process_user,
      process_group: new_resource.process_group,
      worker_processes: new_resource.worker_processes,
      pid: nginx_pid_file,
      worker_connections: new_resource.worker_connections,
      sendfile: new_resource.sendfile,
      tcp_nopush: new_resource.tcp_nopush,
      tcp_nodelay: new_resource.tcp_nodelay,
      keepalive_timeout: new_resource.keepalive_timeout,
      types_hash_max_size: new_resource.types_hash_max_size
    ).merge!(new_resource.conf_variables)
  end
end

action :delete do
  file ::File.join(nginx_dir, 'nginx.conf') do
    action :delete
  end
end
